integration/test_timescaledb.py

```
@@ -306,3 +303,356 @@ async def test_timescaledb_async_emit(timescaledb):
+                max_events=10,
+            ),
+        ]
+    ).run()
+
+    # Test data with extra columns in the event data (should be ignored)
+    test_data = {
+        "time": "18/09/19 01:55:10 UTC+0000",
+        "binary_col": b"test_binary",
+        "int_col": 42,
+        "extra_column_not_in_config": "this_column_is_extra",
+        "another_extra_column": "also_extra",
+    }
+
+    # Should work fine - extra columns in event data are ignored
+    controller.emit(test_data, None)
+    controller.terminate()
+    controller.await_termination()
+
+    # Verify data was inserted correctly for configured columns
+    cursor = connection.cursor()
+    cursor.execute(f"SELECT binary_col, int_col FROM {table_name} ORDER BY int_col DESC LIMIT 1")
+    result = cursor.fetchone()
+    assert result[1] == 42  # int_col
+    # PostgreSQL returns memoryview for binary data, convert to bytes for comparison
+    binary_result = bytes(result[0]) if isinstance(result[0], memoryview) else result[0]
+    assert binary_result == b"test_binary"  # binary_col
+    cursor.close()
+
+
+def test_timescaledb_validation_non_dict_data_type_error(timescaledb):
+    """Test that validation properly rejects non-dictionary data types"""
+    (_, _, _, _, _, _, table_name, dsn_url, _, columns_config) = timescaledb
+
+    time_format = "%d/%m/%y %H:%M:%S UTC%z"
+    controller = build_flow(
+        [
+            SyncEmitSource(),
+            TimescaleDBTarget(
+                dsn=dsn_url,
+                table=table_name,
+                time_col="time",
+                columns=columns_config,
+                time_format=time_format,
+                max_events=10,
+            ),
+        ]
+    ).run()
+
+    # Test with non-dictionary data
+    invalid_data = "this_is_not_a_dictionary"
+
+    # Should raise TypeError for non-dictionary data (the error occurs in the parent Writer class)
+    with pytest.raises(TypeError, match=r"string indices must be integers"):
```

If we're already testing it, maybe we should raise a better error?

---

will fix

---

I think you might have missed this one though since this line wasn't updated.

---

Actually the error is raise by Writer::_event_to_writer_entry(), so not sure what to test there

---

What I see in `Writer::_event_to_writer_entry()` is that it's supposed to raise `TypeError("Writer supports only events of type dict or list.")`, so why are we getting `string indices must be integers`?

---

The source of the "string indices must be integers" error is in storey/targets.py:204 in the _get_event_time() method:

  event_time = event.body[time_field]

  Where:
  - event.body = "this_is_not_a_dictionary" (string)
  - time_field = "time" (string)
  - Attempting: "this_is_not_a_dictionary"["time"] → TypeError: string indices must be integers

  Complete Call Stack Analysis

  Direct method call path:
  1. target._event_to_batch_entry(event)
  2. self._event_to_writer_entry(event)
  3. Validates isinstance(data, dict) → False
  4. Checks elif self._columns: → True
  5. Raises: TypeError: Writer supports only events of type dict or list.

  Flow processing path:
  1. Flow batching framework calls _get_event_time() first
  2. _get_event_time() tries event.body[time_field]
  3. "string_data"["time"] → TypeError: string indices must be integers
  4. Never reaches the proper validation!

---

Cool... Can you fix it? 😄

Even if it's pre-existing code, we shouldn't be adding tests that expect the wrong thing to happen.

---

will remove the test